<?xml version='1.0'?>

<robot name="rodrone" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <link name='center_frame'>
    <pose>0 0 0.1 0 0 0</pose>

    <inertial>
      <mass value="15.0"/>
      <origin xyz="0.0 0 0.1" rpy=" 0 0 0"/>
      <inertia
          ixx="0.1" ixy="0" ixz="0"
          iyy="0.1" iyz="0"
          izz="0.1"
      />
    </inertial>

    <collision name='collision'>
      <geometry>
        <cylinder length="0.02" radius="0.1"/>
      </geometry>
    </collision>

    <visual name='center_frame_visual'>
      <origin xyz="0 0 0" rpy=" 0 0 0"/>
      <geometry>
        <cylinder length="0.02" radius="0.1"/>
      </geometry>
    </visual>
  </link>

  <link name="pos_pos_arm">
    <collision name="collision">
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.04 0.02"/>
      </geometry>
    </collision>
    <visual name="pos_pos_arm_visual">
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.04 0.02"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <mass value="5"/>
      <inertia
        ixx=".1" ixy="0.0" ixz="0.0"
        iyy=".1" iyz="0.0"
        izz=".1"/>
    </inertial>
  </link>

  <link name="neg_pos_arm">
    <collision name="collision">
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.04 0.02"/>
      </geometry>
    </collision>
    <visual name="neg_pos_arm_visual">
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.04 0.02"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <mass value="5"/>
      <inertia
        ixx=".1" ixy="0.0" ixz="0.0"
        iyy=".1" iyz="0.0"
        izz=".1"/>
    </inertial>
  </link>

  <link name="neg_neg_arm">
    <collision name="collision">
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.04 0.02"/>
      </geometry>
    </collision>
    <visual name="neg_neg_arm_visual">
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.04 0.02"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <mass value="5"/>
      <inertia
        ixx=".1" ixy="0.0" ixz="0.0"
        iyy=".1" iyz="0.0"
        izz=".1"/>
    </inertial>
  </link>

  <link name="pos_neg_arm">
    <collision name="collision">
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.04 0.02"/>
      </geometry>
    </collision>
    <visual name="pos_neg_arm_visual">
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.04 0.02"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <mass value="5"/>
      <inertia
        ixx=".1" ixy="0.0" ixz="0.0"
        iyy=".1" iyz="0.0"
        izz=".1"/>
    </inertial>
  </link>

  <link name="pos_pos_rotor">
    <collision name="collision">
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.05"/>
      </geometry>
    </collision>
    <visual name="pos_pos_rotor_visual">
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.1 0.1 0.01"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <mass value="5"/>
      <inertia
        ixx=".1" ixy="0.0" ixz="0.0"
        iyy=".1" iyz="0.0"
        izz=".1"/>
    </inertial>
  </link>

  <link name="neg_pos_rotor">
    <collision name="collision">
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.05"/>
      </geometry>
    </collision>
    <visual name="neg_pos_rotor_visual">
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.1 0.1 0.01"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <mass value="5"/>
      <inertia
        ixx=".1" ixy="0.0" ixz="0.0"
        iyy=".1" iyz="0.0"
        izz=".1"/>
    </inertial>
  </link>

  <link name="neg_neg_rotor">
    <collision name="collision">
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.05"/>
      </geometry>
    </collision>
    <visual name="neg_neg_rotor_visual">
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.1 0.1 0.01"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <mass value="5"/>
      <inertia
        ixx=".1" ixy="0.0" ixz="0.0"
        iyy=".1" iyz="0.0"
        izz=".1"/>
    </inertial>
  </link>

  <link name="pos_neg_rotor">
    <collision name="collision">
      <origin xyz="0 0 0" rpy="0 -1.57079632679 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.05"/>
      </geometry>
    </collision>
    <visual name="pos_neg_rotor_visual">
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.1 0.1 0.01"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <mass value="5"/>
      <inertia
        ixx=".1" ixy="0.0" ixz="0.0"
        iyy=".1" iyz="0.0"
        izz=".1"/>
    </inertial>
  </link>


  <joint type="fixed" name="pos_pos_arm_joint">
    <origin xyz="0.0708 0.0708 0" rpy="0 0 0.78539816339"/>
    <child link="pos_pos_arm"/>
    <parent link="center_frame"/>
  </joint>

  <joint type="fixed" name="pos_neg_arm_joint">
    <origin xyz="-0.0708 0.0708 0" rpy="0 0 2.35619449019"/>
    <child link="pos_neg_arm"/>
    <parent link="center_frame"/>
  </joint>

  <joint type="fixed" name="neg_neg_arm_joint">
    <origin xyz="-0.0708 -0.0708 0" rpy="0 0 -2.35619449019"/>
    <child link="neg_neg_arm"/>
    <parent link="center_frame"/>
  </joint>

  <joint type="fixed" name="neg_pos_arm_joint">
    <origin xyz="0.0708 -0.0708 0" rpy="0 0 -0.78539816339"/>
    <child link="neg_pos_arm"/>
    <parent link="center_frame"/>
  </joint>

  <joint type="continuous" name="pos_pos_rotor_joint">
    <origin xyz="0.17 0 0.02" rpy="0 0 0"/>
    <child link="pos_pos_rotor"/>
    <parent link="pos_pos_arm"/>
    <axis xyz="0 0 1"/> 
  </joint>

  <joint type="continuous" name="neg_pos_rotor_joint">
    <origin xyz="0.17 0 0.02" rpy="0 0 0"/>
    <child link="neg_pos_rotor"/>
    <parent link="neg_pos_arm"/>
    <axis xyz="0 0 1"/> 
  </joint>

  <joint type="continuous" name="neg_neg_rotor_joint">
    <origin xyz="0.17 0 0.02" rpy="0 0 0"/>
    <child link="neg_neg_rotor"/>
    <parent link="neg_neg_arm"/>
    <axis xyz="0 0 1"/> 
  </joint>

  <joint type="continuous" name="pos_neg_rotor_joint">
    <origin xyz="0.17 0 0.02" rpy="0 0 0"/>
    <child link="pos_neg_rotor"/>
    <parent link="pos_neg_arm"/>
    <axis xyz="0 0 1"/> 
  </joint>

  <transmission name="transmission_pos_pos_rotor">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="pos_pos_rotor_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_pos_pos_rotor">
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="transmission_pos_neg_rotor">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="pos_neg_rotor_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_pos_neg_rotor">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="transmission_neg_neg_rotor">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="neg_neg_rotor_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_neg_neg_rotor">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="transmission_neg_pos_rotor">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="neg_pos_rotor_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_neg_pos_rotor">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/rodrone</robotNamespace>
      <legacyModeNS>true</legacyModeNS>
    </plugin>
  </gazebo>

  <gazebo>
    <plugin name="rodrone_plugin" filename="librodrone_plugin.so"/>
  </gazebo>
</robot>

